crates/log-server/src/rocksdb_logstore/keys.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2024 -  Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+// todo: remove after scaffolding is complete
+#![allow(unused)]
+
+use std::mem::size_of;
+
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+
+use restate_bifrost::loglet::LogletOffset;
+use restate_types::logs::SequenceNumber;
+use restate_types::replicated_loglet::ReplicatedLogletId;
+
+const DATA_KEY_PREFIX: u8 = b'd';
+const METADATA_KEY_PREFIX: u8 = b'm';
+pub(crate) const MARKER_KEY: &[u8] = b"storage-marker";
+
+pub(crate) const DATA_KEY_PREFIX_LENGTH: usize = RecordKey::prefix_size();
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct RecordKey {
+    pub loglet_id: ReplicatedLogletId,
+    pub offset: LogletOffset,
+}
+
+impl RecordKey {
+    pub fn new(loglet_id: ReplicatedLogletId, offset: LogletOffset) -> Self {
+        Self { loglet_id, offset }
+    }
+
+    pub fn upper_bound(loglet_id: ReplicatedLogletId) -> Self {
+        Self {
+            loglet_id,
+            offset: LogletOffset::MAX,
+        }
+    }
+
+    pub const fn serialized_size() -> usize {
+        size_of::<Self>() + 1
+    }
+
+    pub const fn prefix_size() -> usize {
+        size_of::<ReplicatedLogletId>() + 1
+    }
+
+    pub fn to_bytes(self) -> BytesMut {
+        let mut buf = BytesMut::with_capacity(Self::serialized_size());
+        self.encode(&mut buf);
+        buf
+    }
+
+    pub fn encode_and_split(self, buf: &mut BytesMut) -> BytesMut {
+        self.encode(buf);
+        buf.split()
+    }
+
+    pub fn encode(self, buf: &mut BytesMut) {
+        buf.reserve(Self::serialized_size());
+        buf.put_u8(DATA_KEY_PREFIX);
+        buf.put_u64(*self.loglet_id);
+        // despite offset being 32bit, for backward compatibility and future-proofing, we keep the
+        // storage layer as u64 but we assert that the actual value fits u32 on read.
+        buf.put_u64(self.offset.into());
+    }
```

Comment:
This is a not a platform independent encoding. since it has no regard for endianness .

Reply:
I think it writes the u64 in big endian byte order.